Flatcar integration

heytrav opened the pull request:
* Modify openstack-cluster templates to integrate ignition components for Flatcar
* Smaller changes related to addon defaults

mkjpryor commented:
I don't like how Flatcar feels like a bolt-on here. If we are going to support it, I feel we should refactor the charts so that there is an OS abstraction that Ubuntu and Flatcar both use.

mkjpryor commented:
I think there is a nice way to do it by utilising template definitions and passing around the name of a template to use, but I haven't got it fully figured out yet.

heytrav commented:
> I think there is a nice way to do it by utilising template definitions and passing around the name of a template to use, but I haven't got it fully figured out yet.

I wonder if something similar could be done for #91. Seems like some sort of abstraction for adding patches to the kube-apiserver config could also be useful.

mkjpryor commented:
@heytrav I was thinking the same. Let me have a bit of a think about how it would work.

heytrav commented:
@mkjpryor based the changes you made in #130 it looks like I can just override `controlPlane.kubeadmConfigSpec` and `nodeGroupDefaults.kubeadmConfigSpec` to add the necessary ignition components in a user defined `-f values.yaml`.  I'm still experimenting with it to be sure.

For the purpose of this PR I'm wondering if you want to 

1. Provide any sort of conditional blocks in `charts/openstack-cluster/values.yaml` to switch between cloud-init or ignition based systems OR
2. Leave that out completely and leave it up to the user to override with the ignition components?

For the latter case I would probably just close this PR. We'll just use our own with the ignition setup regardless.

heytrav commented:
I've refactored the original PR to remove most of the changes I had made. In the current state it is possible to build Flatcar clusters with this code and providing customised `values.yaml`.

However as mentioned in #138  it was necessary to change the last command added to `preKubeadmCommands` because it references a folder that does not exist on Flatcar and causes the build to fail. Rather than remove it completely I added a separate definition that uses a different path when `ignitionBasedOS: true`.

To resolve #139 I changed the destination of the additional containerd config to create a new file under `/etc/containerd/config.d/`. Not sure if this is best approach but for now it at least doesn't clobber the containerd configuration

mkjpryor commented on charts/openstack-cluster/templates/_helpers.tpl:
```diff
@@ -203,10 +203,11 @@ files:
-  - path: /etc/containerd/config.toml
+  - path: /etc/containerd/config.d/containerd-certs.toml
     content: |
-      [plugins."io.containerd.grpc.v1.cri".registry]
-      config_path = "/etc/containerd/certs.d"
+      [plugins]
+        [plugins."io.containerd.grpc.v1.cri".registry]
+        config_path = "/etc/containerd/certs.d"
```
We actually tried this, but it doesn't work because of this:

https://github.com/containerd/containerd/issues/5837

That is why we used `append: true` here, which clearly doesn't work in Ignition.

So this change will need to be reverted and we will need to think of another solution. It would be nice to be able to do this in Flatcar also (i.e. support registry mirrors).

mkjpryor commented on charts/openstack-cluster/templates/control-plane/kubeadm-control-plane.yaml:
```diff
@@ -21,6 +21,22 @@ preKubeadmCommands:
 {{- end }}
 {{- end }}
 
+{{- define "openstack-cluster.controlplane.kubeadmConfigSpec.ignitionKubeProxyConfiguration" -}}
+{{- with .kubeProxyConfiguration }}
+files:
+  - path: /etc/kube-proxy-configuration.yaml
+    content: |
+      ---
+      apiVersion: kubeproxy.config.k8s.io/v1alpha1
+      kind: KubeProxyConfiguration
+      {{- toYaml . | nindent 6 }}
+    owner: root:root
+    permissions: "0644"
+preKubeadmCommands:
+  - cat /etc/kube-proxy-configuration.yaml >> /run/kubeadm.yml
+{{- end }}
+{{- end }}
```
Can this be done by just gating the path and command on `.Values.ignitionBasedOS`? I'm not that keen on the duplication.

heytrav replied:
Inside the block I was unable to access the  `.Values.ignitionBasedOS` flag. The helm chart would fail to render because the `.ignitionBasedOS` variable didn't exist. Maybe that happens because it is in the `.kubeProxyConfiguration` context?  I'm not sure. 

Perhaps there's a trick to getting the root level objects

mkjpryor commented on charts/openstack-cluster/templates/control-plane/kubeadm-control-plane.yaml:
```diff
@@ -47,6 +63,22 @@ spec:
     nodeDrainTimeout: {{ .Values.controlPlane.nodeDrainTimeout }}
     nodeVolumeDetachTimeout: {{ .Values.controlPlane.nodeVolumeDetachTimeout }}
     nodeDeletionTimeout: {{ .Values.controlPlane.nodeDeletionTimeout }}
+  {{- if .Values.ignitionBasedOS }}
+  kubeadmConfigSpec: {{
+    omit
+      (
+        list
+          (include "openstack-cluster.controlplane.kubeadmConfigSpec.nodeLabels" . | fromYaml)
+          (include "openstack-cluster.kubeadmConfigSpec" (list . .Values.controlPlane.kubeadmConfigSpec) | fromYaml) 
+          (include "openstack-cluster.controlplane.kubeadmConfigSpec.ignitionKubeProxyConfiguration" .Values.controlPlane.kubeadmConfigSpec | fromYaml) |
+        include "openstack-cluster.mergeConcatMany" |
+        fromYaml
+      )
+      "kubeProxyConfiguration" |
+    toYaml |
+    nindent 4
+  }}
+  {{- else }}
```
Which would mean this is not needed. We would need to modify the call below to pass in the whole context.

heytrav commented:
> @heytrav
> 
> I think in the short- to medium-term it would be nice to have the Flatcar support be native to the charts, so that we only require the user to flip a value in the charts between `flatcar` and `ubuntu`. However if you think this is the minimum set of changes that enable the use of Flatcar and you don't have the time or inclination to look at the required changes for that, I am happy to merge (subject to a full review and passing the test suite).
> 
> Let me know what you prefer.

I actually would like to have ignition based OS (i.e. flatcar) be completely handled by the chart as you suggest. However I wasn't sure how you preferred this to work. 

I will need to experiment a bit to add this into the chart. If you have any suggestions, I'm happy to follow your lead.

heytrav commented:
Most recent commit works for me locally (I can build a Flatcar cluster).

@mkjpryor please let me know if this is more in line with what you were thinking.

mkjpryor commented on charts/openstack-cluster/templates/_helpers.tpl:
```diff
@@ -203,13 +203,15 @@ files:
       # This file is created by the capi-helm-chart to ensure that its parent directory exists
     owner: root:root
     permissions: "0644"
+{{- if not (.Values.ignitionBasedOS | default false) }}
```
I would prefer we just set `ignitionBasedOS: false` in the chart `values.yaml`, then we don't need the `default` here.

mkjpryor replied:
Similarly elsewhere

mkjpryor commented on charts/openstack-cluster/values.yaml:
```diff
@@ -119,6 +119,39 @@ apiServer:
   # The port to use for the API server
   port: 6443
 
+
+# Set ignition based OS
+# ignitionBasedOS:
+# Ignition Based OS specific configuration. 
+ignitionKubeadmConfigSpec:
+  initConfiguration:
+    nodeRegistration:
+      name: ${COREOS_OPENSTACK_HOSTNAME}
+  joinConfiguration:
+    nodeRegistration:
+      name: ${COREOS_OPENSTACK_HOSTNAME}
+  preKubeadmCommands:
+    - export COREOS_OPENSTACK_HOSTNAME=${COREOS_OPENSTACK_HOSTNAME%.*}
+    - envsubst < /etc/kubeadm.yml > /etc/kubeadm.yml.tmp
+    - mv /etc/kubeadm.yml.tmp /etc/kubeadm.yml
+  format: ignition
+  ignition:
+    containerLinuxConfig:
+      additionalConfig: |
+        systemd:
+          units:
+          - name: coreos-metadata-sshkeys@.service
+            enabled: true
+          - name: kubeadm.service
+            enabled: true
+            dropins:
+            - name: 10-flatcar.conf
+              contents: |
+                [Unit]
+                Requires=containerd.service coreos-metadata.service
+                After=containerd.service coreos-metadata.service
+                [Service]
+                EnvironmentFile=/run/metadata/flatcar
```
I was actually thinking more like all this stuff being in the `openstack-cluster.ignitionKubeadmConfigSpec`, so that we don't even have a separate variable here. This would be more consistent with the existing setup. Maybe that template should be called `openstack-cluster.flatcarKubeadmConfigSpec` in case there are other Ignition-based OSs that have different requirements to Flatcar, and instead of `.Values.ignitionBasedOS` we can have `.Values.osName` that defaults to `ubuntu`.

It might also involve moving anything that is Ubuntu-specific from `values.yaml` to a corresponding `openstack-cluster.ubuntuKubeadmConfigSpec` template.

The user can still customise the kubeadmConfigSpec for either Flatcar or Ubuntu via the existing hooks.

heytrav replied:
Should we call it `osDistro` to align with the glance image convention?

heytrav commented:
> @heytrav
> 
> I added my thoughts in the form of review comments. Since we are adding support for a second OS, I would like to make sure that we can actually support N OSs going forward, just in case, so the suggestions I made reflect that.

pushed up some changes

mkjpryor reviewed (approved):
LGTM
